Kubeadm upload and fetch of kubeam config v1alpha3

fabriziopandini opened the pull request:
**What this PR does / why we need it**:
This PR implements upload and fetch of kubeam config v1alpha3 from cluster.

More in detail:
In upload, `kubeadm-config` gets
- `ClusterConfiguration` (without components config which are already stored in separated ConfigMaps)
- `ClusterStatus`(initialised or updated with the API endpoint of the current node)

During fetch `InitConfiguration` is composed with:
- `ClusterConfiguration` from `kubeadm-config`
- The `APIEndpoint` of the current node from `ClusterStatus` in `kubeadm-config`
- Component configs from corresponding ConfigMaps

**Which issue(s) this PR fixes** :
refs https://github.com/kubernetes/kubeadm/issues/911, refs https://github.com/kubernetes/kubeadm/issues/963

**Special notes for your reviewer**:
In order to implement this it was necessary to extend current component config management with a new GetFromConfigMap operation. This is implemented in a separated commit "
implement component configs GetFromConfigMap".
The real change build on this (commi "upload and fetch kubeadm v1alpha3")

**Release note**:
```release-note
NONE
```

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/sig cluster-lifecycle
/area kubeadm
/kind enhancement
/assign @luxas
/assign @timothysc
/cc @chuckha @rosti @neolit123 @liztio

neolit123 commented on cmd/kubeadm/app/componentconfigs/config_test.go:
```diff
@@ -0,0 +1,119 @@
+	// Create temp folder for the test case
+	tmpdir, err := ioutil.TempDir("", "")
+	if err != nil {
+		t.Fatalf("Couldn't create tmpdir")
```
`Couldn't` -> `couldn't`

neolit123 commented on cmd/kubeadm/app/componentconfigs/config_test.go:
```diff
@@ -0,0 +1,119 @@
+			}
+
+			if obj == nil {
+				t.Errorf("Unexpected nil return value")
```
`Unexpected` -> `unexpected`

neolit123 commented on cmd/kubeadm/app/componentconfigs/config_test.go:
```diff
@@ -0,0 +1,119 @@
+			k8sVersion := version.MustParseGeneric("v1.12.0")
+			obj, err := r.GetFromConfigMap(client, k8sVersion)
+			if err != nil {
+				t.Errorf("Error getting ConfigMap: %v", err)
```
`Error getting` -> `error getting`

neolit123 commented on cmd/kubeadm/app/componentconfigs/config_test.go:
```diff
@@ -0,0 +1,119 @@
+
+			err := rt.configMap.create(client)
+			if err != nil {
+				t.Errorf("Couldn't create ConfigMap %s", rt.configMap.name)
```
`Couldn't ` -> `couldn't `

neolit123 reviewed (commented):
thanks @fabriziopandini 
added minor comments but also commented on the usage of ChooseBindAddress that we talked about last time. we might need something more unique here, unless i'm mistaken.

rosti commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -23,54 +23,140 @@ import (
 )
 
 // FetchConfigFromFileOrCluster fetches configuration required for upgrading your cluster from a file (which has precedence) or a ConfigMap in the cluster
-// TODO: This func should be renamed FetchClusterConfigFromFileOrCluster, and return a ClusterConfiguration instead of an InitConfiguration
```
We may need to keep an updated version of this TODO for the future.
Ideally we should not be forced to have InitConfiguration for anything else than `kubeadm init`. I am not sure, that ClusterConfiguration is the answer too (since APIEndpoint is not part of it).
It may be something else, but we may just keep some TODO here, just to mark our intent in reducing the InitConfiguration usages in the long run.

fabriziopandini replied:
Here I removed the comment about `ClusterConfiguration` because we know this isn't the right approach. 

For the long run I agree 100% with your intent of reducing the InitConfiguration usages, but better track this with an issue than TODOs in the code.
If you can kindly open the issue this will be great 😉

rosti replied:
Filed kubernetes/kubeadm#1084

rosti reviewed (commented):
@fabriziopandini Thanks for the effort! Has a few minor nits, but otherwise it LGTM.

fabriziopandini commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -373,6 +376,46 @@ type AuditPolicyConfiguration struct {
+func GetMachineID() (string, error) {
+	// the IP of the host default network interface as a machine unique id
+	ip, err := netutil.ChooseBindAddress(nil)
+	if err != nil {
```
@neolit123 how frequent is this case?  
However, If there is a library already vendored in kubernetes for getting UUIDs I can try to switch...

timothysc replied:
I think generating some UUID is pretty useful when folks are changing nodes in a virutual infra will be useful.

neolit123 replied:
i can't find a vendored "system" uuid lib in k8s - i.e. a library for returning the uuid of the machine.

there is this, but this library is just for generating a RFC compliant random uuid:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/uuid/uuid.go

we need something like:
https://github.com/denisbrodbeck/machineid
the the methods that the library uses are outlined here:
https://github.com/denisbrodbeck/machineid#snippets

rosti replied:
Meanwhile, in our own documentation:
https://kubernetes.io/docs/setup/independent/install-kubeadm/#verify-the-mac-address-and-product-uuid-are-unique-for-every-node

neolit123 replied:
@fabriziopandini we talked about this at the kubeadm office hours today.

Jason reminded us that the node name should be unique for a cluster to work anyway so we can use that as the key for the status map:

from @detiber 
>More specifically, this should be "node name" being either the user-supplied override in the kubeadm config or the default detected node name (generally the hostname here)

timothysc commented on cmd/kubeadm/app/componentconfigs/config.go:
```diff
@@ -0,0 +1,80 @@
+
+// HasKubeletConfigMap returns true if the Kubelet is expected to have a config map stored in the cluster
+func HasKubeletConfigMap(cfg *kubeadmapi.ClusterConfiguration) bool {
+	return true
```
umm

fabriziopandini replied:
Forget it, now removed

timothysc commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -187,14 +187,25 @@ const (
+	KubeProxyConfigMap = "kube-proxy"
+
+	// KubeProxyConfigMapKey specifies in what ConfigMap key the component config of kube-proxy should be stored
+	KubeProxyConfigMapKey = "config.conf"
```
?

fabriziopandini replied:
This is the key name used in kube-proxy manifest.. just refactored in a constant because it is now used in different places.

timothysc reviewed (commented):
I only did a quick scan, let me know when you're ready and I'll go through this in detail.

rosti commented on cmd/kubeadm/app/phases/addons/proxy/manifests.go:
```diff
@@ -79,7 +79,7 @@ spec:
         command:
         - /usr/local/bin/kube-proxy
-        - --config=/var/lib/kube-proxy/config.conf
+        - --config=/var/lib/kube-proxy/{{ .ProxyConfigMapKey }}s
```
Is the trailing `s` desired?

fabriziopandini replied:
No, now fixed, thanks

fabriziopandini commented:
@timothysc 
1) Implemented the logic for getting the node name from the `kubelet.conf`. However it was necessary to get the user name from x509 certificate and manage two cases:
   - `kubelet.conf` generated by kubeadm 
   - `kubelet.conf` generated by TLSBootstrap.
2) Used the node name as a key for the `ClusterConfig.APIEndpoints` map (insterad of the netutil.ChooseBindAddress or hostname)
3) Implemented a lot of unit tests for all the upload /fetch config jiggery
4) Fixed `kubeadm join --control-plane`, `kubeadm upgrade` and `kubeadm upgrade node --control-plane` workflows with the new upload /fetch config jiggery removing some TODOs left in previous PRs
5) completed first rounds of manual testing:
    - V1.11.2 --> v1.12.0
    - V1.12.0 with two masters --> v1.12.1

timothysc commented on cmd/kubeadm/app/componentconfigs/config.go:
```diff
@@ -0,0 +1,68 @@
+
+	// Decodes the Config map into the internal component config
+	obj := &kubeletconfig.KubeletConfiguration{}
+	err = unmarshalObject(obj, []byte(kubeletCfg.Data[kubeadmconstants.KubeletBaseConfigurationConfigMapKey]))
```
should you index check 1st otherwise you could SEGV here.  .Data[kubeadmconstants.KubeletBaseConfigurationConfigMapKey]

fabriziopandini replied:
fixed + added UT

timothysc commented on cmd/kubeadm/app/componentconfigs/config.go:
```diff
@@ -0,0 +1,68 @@
+
+	// Decodes the Config map into the internal component config
+	obj := &kubeproxyconfig.KubeProxyConfiguration{}
+	err = unmarshalObject(obj, []byte(kubeletCfg.Data[kubeadmconstants.KubeProxyConfigMapKey]))
```
Same comment.

fabriziopandini replied:
same

timothysc commented on cmd/kubeadm/app/componentconfigs/registrations.go:
```diff
@@ -60,13 +64,20 @@ func (r Registration) Unmarshal(fileContent []byte) (runtime.Object, error) {
 	return obj, nil
 }
 
+func unmarshalObject(obj runtime.Object, fileContent []byte) error {
```
I swear these functions exist elsewhere.

fabriziopandini replied:
This function uses a scheme with component configs api registered.
The other `unmarshal` func uses a scheme with only kubeadm types registered

timothysc commented on cmd/kubeadm/app/phases/addons/proxy/proxy.go:
```diff
@@ -144,5 +151,39 @@ func createClusterRoleBindings(client clientset.Interface) error {
+	}
+
+	// Create a role for granting read only access to the kube-proxy component config ConfigMap
+	if err := apiclient.CreateOrUpdateRole(client, &rbac.Role{
```
Do we have tests to validate the updated RBAC rules?

fabriziopandini replied:
The only test on RBAC I'm aware of are in `test\e2e_kubeadm` but I'm not sure if those test are used in test grid batches...

timothysc commented on cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go:
```diff
@@ -39,20 +42,35 @@ const (
+
+	// Prepare the ClusterStatus for upload
+	// Gets the current cluster status
+	clusterStatus, err := getClusterStatus(client)
```
I think we are ok for now, but we should add a //TODO use configmap locks on this object on the get before the update.  We may even bug fix that after the freeze 

/cc @liztio

fabriziopandini replied:
done

timothysc reviewed (changes requested):
1st pass run through.

neolit123 commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -17,60 +17,243 @@ limitations under the License.
+			return "", err
+		}
+	} else {
+		return "", fmt.Errorf("Invalid kubelet.conf. X509 certificate expected")
```
nit / optional: `errors.New()` instead of `fmt.Errorf()`

neolit123 commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -17,60 +17,243 @@ limitations under the License.
+
+	// We are only putting one certificate in the certificate pem file, so it's safe to just pick the first one
+	// TODO: Support multiple certs here in order to be able to rotate certs
+	cert := certs[0]
```
do we have a tracking issue?

neolit123 reviewed (approved):
LGTM

rosti reviewed (approved):
@fabriziopandini at first pass this LGTM, has some tiny nits though

fabriziopandini committed:
upload and fetch of kubeam config v1alpha3 from cluster

fabriziopandini committed:
use new kubeadm-config in kubeadm join control-plane & kubeadm upgrade/upgrade node

fabriziopandini committed:
autogenerated

timothysc reviewed (approved):
/lgtm 
/approve 

We're too late for nits, we'll have to beat on this during freeze.

timothysc commented:
/test pull-kubernetes-integration
